File: pattern/prefix_sum/525_Contiguous_Array/Solution.java
class Solution {
    public int findMaxLength(int[] nums) {
        int[] preSum = new int[nums.length + 1];
        Map<Integer, Integer> sumToIdx = new HashMap<>();
        preSum[0]=0;
        sumToIdx.put(preSum[0], 0);
        for(int i =0 ; i < nums.length ; i++){
            if(nums[i]==0){
                nums[i]= -1;
            }
        }
       int max = 0;
        for(int i = 1 ; i < nums.length+1 ; i++){
            preSum[i]=preSum[i-1]+nums[i-1];
            if(sumToIdx.containsKey(preSum[i])){
                int idx = sumToIdx.get(preSum[i]);
                max = Math.max(max , i - idx);
            }
            else{
                sumToIdx.put(preSum[i],i);
            }
        }

        return max;
    }
}

File: pattern/prefix_sum/525_Contiguous_Array/Solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function(nums) {
    let sum = 0;
    let presum = new Map();
    let res = 0;
    presum.set(0,-1)
    for(let i = 0 ; i < nums.length ; i++){
        if(nums[i]==0){
            nums[i]=-1;
        }
        sum+=nums[i];
        if(presum.has(sum)){
            res = Math.max(res,i - presum.get(sum))
        }
        else{
            presum.set(sum,i)
        }
    }
    return res;
};
